### PR TITLE
Plugin ready for code review

### PR DESCRIPTION
**Pull request by berleant**
- [ ] UI
  - [ ] modify sign-up modal
    - [x] Handlebars template
      - [x] make group full names accessible to UI via ps store
    - [x] connector
    - [x] Ember component
    - [ ] better user feedback on collectives, debt amount, phone fields -- might skip this if it takes too long
- [x] platform
  - [x] validate input
  - [x] handle rails auto"magically" doing a bad thing (list vs string)
  - [x] persist user data
    - [x] js
    - [x] ruby
      - [x] allow custom fields to pass through
      - [x] make sure rspec for UserUpdater.update still works
      - [x] handle errors in event hooks
  - [x] sign user up for the correct groups on sign up
    - [x] ensure group member counts correct
  - [x] sign user up for the correct groups if they change collectives
- [x] testing
  - [x] qunit acceptance test for sign-up modal behavior

### CODE REVIEW

File: assets/javascripts/discourse/templates/components/tds-new-user.hbs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+<tr class="input">
+  <td></td>
+  <td class="collective-selectors">
+    <!-- one category per collective -->
+    {{#each site.categories as |c|}}
```
Comment (sarayourfriend) [inline assets/javascripts/discourse/templates/components/tds-new-user.hbs]: We probably need to filter these a little further, right? Idk if we can be certain the the _only_ categories will be the collective ones.

**Comment by sarayourfriend** [comment]
I don't really know too much about discourse plugins at this point but other than the one question LGTM! :+1:

File: assets/javascripts/discourse/templates/components/tds-new-user.hbs (hunk 1 above)
Comment (berleant) [reply assets/javascripts/discourse/templates/components/tds-new-user.hbs]: Ah, I thought that was a guarantee. I'll look into how to do that.

Comment (sarayourfriend) [reply assets/javascripts/discourse/templates/components/tds-new-user.hbs]: There might be ones for the dispute admins to talk and stuff like that? I feel like we may have mentioned something along those lines.

File: assets/javascripts/discourse/components/tds-new-user.js.es6 [hunk 2]
```diff
@@ -1,17 +1,18 @@
...
     collectiveSelected(event) {
-      const solidarityBloc = "Solidarity Bloc"
 
       const elements = $(".collective-selectors input:checked").toArray()
       const collectives = elements.map(element => element.attributes.value.value)
       
-      Ember.set(this, 'isInSolidarity', collectives.includes(solidarityBloc))
-      Ember.set(this, 'isDebtor', collectives.filter(x => x != solidarityBloc).length > 0)
+      Ember.set(this, 'isInSolidarity', collectives.includes(solidarity))
+      Ember.set(this, 'isDebtor', collectives.filter(x => x != solidarity).length > 0)
```
Comment (sarayourfriend) [inline assets/javascripts/discourse/components/tds-new-user.js.es6]: You can use Array.prototype.find here instead.

Comment (berleant) [reply assets/javascripts/discourse/components/tds-new-user.js.es6]: Cool, I didn't know that. I guess there is a small performance benefit (?) to using find, but I think I'll leave it as is because I think it's more readable this way.

Comment (berleant) [reply assets/javascripts/discourse/components/tds-new-user.js.es6]: Or maybe it isn't. I can't decide.

**Comment by orlando** [comment]
@shoshber Do we need to create UI for this? Discourse has a way to render these in the signup form, but not enable that option by default. Not sure if we can just use that.

(http://localhost:3000/admin/customize/user_fields)
image

**Comment by berleant** [comment]
Yes, the built-in user field UI is too brittle. Not customizable at all.

File: plugin.rb [hunk 3]
```diff
@@ -48,6 +35,19 @@ def set_up_event_triggers()
 end
 
 def assign_user_groups(user)
+  # each collective has a correspnding user group (value) and category (key)
+  # see seed.js debtsyndicate/discourse-seed
+  collectives = {
+    'For Profit Colleges Collective': 'for-profit-colleges',
+    'Student Debt Collective': 'student0debt',
```
Comment (sarayourfriend) [inline plugin.rb]: Looks like a typo in `student-debt`

Comment (berleant) [reply plugin.rb]: yep thanks

**Comment by berleant** [comment]
There is an edge case where the account creation succeeds but assigning a collective fails that I did not handle.

**Comment by orlando** [comment]
@shoshber why that happens?

**Comment by berleant** [comment]
@orlando we talked about this offline but just to close the loop here, it should never happen, but it technically could if the server crashed or something of that nature.

File: prettier.config.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+module.exports = {
```
Comment (sarayourfriend) [inline prettier.config.js:1]: Would it be possible to use [prettier's overrides](https://prettier.io/docs/en/configuration.html#configuration-overrides) to get it to apply to the `.js.es6` files?

File: assets/javascripts/discourse/mixins/debt-amount-validation.js.es6 [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import InputValidation from 'discourse/models/input-validation';
+import { default as computed } from 'ember-addons/ember-computed-decorators';
```
Comment (sarayourfriend) [inline assets/javascripts/discourse/mixins/debt-amount-validation.js.es6:2]: To import the default you can just do `import computed from 'ember-addons/ember-computed-decorators'` unless there's something specifically different for whatever reason.

File: assets/javascripts/discourse/initializers/tdc-user-model.js.es6 [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        challenge: attrs.accountChallenge,
+      }
+
+      data = this.uglyHackPartTwo(data, attrs)
+
+      return ajax(userPath(), { data, type: 'POST'})
+    },
+
+    uglyHackPartTwo(data, attrs) {
+        // see uglyHackPartOne, in tdc-create-account-controller
```
Comment (sarayourfriend) [inline assets/javascripts/discourse/initializers/tdc-user-model.js.es6]: :+1:

File: assets/javascripts/discourse/initializers/tdc-user-model.js.es6 [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import { withPluginApi } from 'discourse/lib/plugin-api';
+import InputValidation from 'discourse/models/input-validation';
+import createAccountController from 'discourse/controllers/create-account';
+import User from 'discourse/models/user'
+
+import DebtAmountValidation from "../mixins/debt-amount-validation";
+
+function initializeTdcUser(api) {
+  // use this when/if my PR is approved
+  //api.modifyClass('model:user', {
```
Comment (sarayourfriend) [inline assets/javascripts/discourse/initializers/tdc-user-model.js.es6]: Do we still need to do this?

File: assets/javascripts/discourse/templates/components/tdc-sign-up.hbs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    {{input value=store.customFields.debt_amount id="debt-amount" name="debt-amount" required=isDebtor}}
+ <!--   &nbsp;{{input-tip validation=debtAmountValidation id="debt-amount-validation"}} -->
+  </div>
+</div>
+{{/if}}
+
+{{#if isInSolidarity }}
+<tr class="input">
+  <td class="label">
+    <label for="solidarity-how-can-you-help">How can you help the debt collective?</label>
```
Comment (sarayourfriend) [inline assets/javascripts/discourse/templates/components/tdc-sign-up.hbs:26]: Let's get this merged as soon as you can, but if you could create a ticket for @zakgreene to come up with the right copy for these questions that would be sweet 👌

File: package.json [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  "devDependencies": {
+    "husky": "^0.14.3",
+    "lint-staged": "^7.0.0",
+    "prettier": "^1.11.1"
+  },
+  "scripts": {
+    "precommit": "lint-staged"
+  },
+  "lint-staged": {
+    "*.js": ["prettier --write", "git add"]
```
Comment (sarayourfriend) [inline package.json]: Oh, you'll also want to add the js.es6 files to this configuration

**Review by sarayourfriend — APPROVED** [review]
Dang! So much work. LGTM though!

File: assets/javascripts/discourse/initializers/tdc-user-model.js.es6 (hunk 7 above)
Comment (berleant) [reply assets/javascripts/discourse/initializers/tdc-user-model.js.es6]: I'm not sure I understand what you're asking. Can you rephrase or expand?

File: assets/javascripts/discourse/mixins/debt-amount-validation.js.es6 (hunk 5 above)
Comment (berleant) [reply assets/javascripts/discourse/mixins/debt-amount-validation.js.es6:2]: 👍

File: assets/javascripts/discourse/templates/components/tdc-sign-up.hbs (hunk 8 above)
Comment (berleant) [reply assets/javascripts/discourse/templates/components/tdc-sign-up.hbs:26]: https://github.com/debtcollective/parent/issues/133

File: package.json (hunk 9 above)
Comment (berleant) [reply package.json]: 👍

File: prettier.config.js (hunk 4 above)
Comment (berleant) [reply prettier.config.js:1]: I think since I modified `package.json` that this isn't necessary.

File: assets/javascripts/discourse/initializers/tdc-user-model.js.es6 (hunk 7 above)
Comment (sarayourfriend) [reply assets/javascripts/discourse/initializers/tdc-user-model.js.es6]: The comment on 14 says to use the commented out part of 15 when the PR is merged. At least that's what I understood it to mean, maybe it's something else though!  Just wondering what you meant/if we need to do anything with this before merging or if it's just left overs.
